roachtest: collect CPU, Allocs, Mutex profiles after sysbench runs

iskettaneh opened the pull request:
This commit makes sysbench{-settings} runs perform a short sysbench run at the end of the benchmark. During that run, we will collect CPU, Allocs, Mutex profiles. These profiles could feed into PGO, also they can be useful to investigate the performance of a specific run.

Release note: None

Epic: None

iskettaneh committed:
roachtest: collect CPU, Allocs, Mutex profiles after sysbench runs

This commit makes sysbench{-settings} runs perform a short sysbench run
at the end of the benchmark. During that run, we will collect CPU,
Allocs, Mutex profiles. These profiles could feed into PGO, also they
can be useful to investigate the performance of a specific run.

Release note: None

Epic: None

iskettaneh committed:
roachtestutil: add profile collection helper functions

Release note: None

Epic: None

tbg commented on pkg/cmd/roachtest/roachtestutil/profile.go:
```diff
@@ -304,3 +307,144 @@ func MeasureQPS(
+	return getProfileWithTimeout(ctx, logger, client, url, 10*time.Second /* timeout */)
+}
+
+// GetAllocsProfile returns the allocs profile for the given node.
```
Can you add support for delta profiles here, which is what we use almost always? You can add a `seconds int` parameter which is then passed into the URL:

```
url := fmt.Sprintf("https://%s/debug/pprof/allocs?seconds=%s", adminUIAddrs[0], seconds)
```

I believe zero is treated as just returning the absolute profile (i.e. no diffing).

tbg replied:
Actually `duration time.Duration` is probably better, then internally use `int64(duration.Seconds())`.

iskettaneh replied:
Done.
<!-- Reviewable comment -OQJA2Mf-v-j05hS-p01:b-896fix -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachtest/roachtestutil/profile.go:
```diff
@@ -304,3 +307,144 @@ func MeasureQPS(
+}
+
+// GetMutexProfile returns the mutex profile for the given node.
+func GetMutexProfile(
```
ditto about support for diffing.

iskettaneh replied:
Done
<!-- Reviewable comment -OQJKZRNAlQdo-wfebZu:bb6un1j -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachtest/roachtestutil/profile.go:
```diff
@@ -304,3 +307,144 @@ func MeasureQPS(
+// cluster in parallel. It returns a slice of profiles and an error if any of
+// the profiles failed to collect. Each entry in the slices is in the same order
+// as the CRDB nodes in the cluster.
+func GetProfileForAllCRDBNodes(
```
This isn't that useful as a helper, since it forces all (CRDB) nodes but sometimes tests don't have CRDB running on all of them. This would be much more versatile if it accepted an `option.NodeListOption` like most cluster methods.

After my previous suggestion, there's only one method that takes a profile for an individual node. I would make then rename that one to `getProfileSingleNode` (node: not public) and rename this one to `GetProfile` with params `name string` ("cpu", "mutex", ...), a duration, and a node list option. Then you have a single method that fulfills all profiling needs. It returns slices of profiles, but that's fine.

iskettaneh replied:
Done.
<!-- Reviewable comment -OQJGrusDZAK8E2qaQie:b-896fix -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachtest/tests/sysbench.go:
```diff
@@ -259,6 +261,31 @@ func runSysbench(ctx context.Context, t test.Test, c cluster.Cluster, opts sysbe
+		t.Status("running 75 second workload to collect profiles")
+		{
+			// Create a wait group to wait for the profile collection to finish
+			wg := sync.WaitGroup{}
```
I believe there is a better primitive you should use, see here:

https://github.com/cockroachdb/cockroach/blob/7262f5ce8ea9080afc173a61c2b8a614937a3b6f/pkg/cmd/roachtest/tests/decommission.go#L191-L200

and note there's also the `ErrorGroup` variant.

iskettaneh replied:
Done.
<!-- Reviewable comment -OQJKbEdD88A_u40oiNP:b-896fix -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachtest/tests/sysbench.go:
```diff
@@ -596,6 +623,101 @@ func exportSysbenchResults(
+	exportProfile := func(profile *profile.Profile, filename string) {
+		buf := bytes.Buffer{}
+		require.NoError(t, profile.Write(&buf))
+		err = os.WriteFile(filepath.Join(t.ArtifactsDir(), filename), buf.Bytes(), 0644)
```
Let's put these into `filepath.Join(t.PerfArtifactsDir(), "profiles")`. If they got uploaded to our nightly bucket and preserved[^1], this may come in handy - we can go back and track the change in profiles over time if we ever want to do that.
I think all we need for that is this little one-line change in in roachtest_util.sh:

```
      done <<< "$(find . -type f \( -name "${stats_file_name}" -o -name '*.pb.gz' \) | sed 's/^\.\///')")
```

[^1]: https://github.com/cockroachdb/cockroach/blob/7262f5ce8ea9080afc173a61c2b8a614937a3b6f/build/teamcity/util/roachtest_util.sh#L75-L95

tbg commented on pkg/cmd/roachtest/tests/sysbench.go:
```diff
@@ -596,6 +623,101 @@ func exportSysbenchResults(
+	exportProfile(mutexDiffMerged, "mutex-diff-merged.prof")
+	exportProfile(allocsDiffMerged, "allocs-diff-merged.prof")
+	for i := range len(c.CRDBNodes()) {
+		exportProfile(mutexDiffProfiles[i], fmt.Sprintf("mutex-diff-%d.prof", i+1))
```
How about this schema throughout: `n1.mutex30s.pb.gz` (note that `pb.gz` is the canonical extension and I've had to rename files in the past so that tools would accept the file)

tbg reviewed (changes requested):
I have a number of suggested improvements and a few feature requests (sorry 😄) but this is a good start! Thanks for getting this out so quickly.

tbg commented on pkg/cmd/roachtest/tests/sysbench.go:
```diff
@@ -259,6 +261,31 @@ func runSysbench(ctx context.Context, t test.Test, c cluster.Cluster, opts sysbe
+
+			// Wait for 15 seconds to give a chance to the workload to start, and then
+			// collect CPU, mutex diffs, allocs diffs profiles.
+			time.Sleep(15 * time.Second)
```
We have 75s of workload though, and we're only taking a 30s CPU profile, right? I would start the CPU profile at around (shortly after) second 30 then, to get closer to steady state, meaning a 30s sleep here.

iskettaneh committed:
roachtestutil: clean up the profile collection helper functions

Release note: None

Epic: None

iskettaneh commented on pkg/cmd/roachtest/roachtestutil/profile.go:
```diff
@@ -24,6 +25,8 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/google/pprof/profile"
+	"github.com/pkg/errors"
```
Done.
<!-- Reviewable comment -OQJ5I1j0eJAShKK2tq-:b-896fix -->
<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/cmd/roachtest/roachtestutil/profile.go:
```diff
@@ -304,3 +307,144 @@ func MeasureQPS(
+	for {
+		select {
+		case <-timer.C:
+			return nil, errors.Wrapf(latestError, "failed to get profile from %s after %s", url, timeout)
```
Good catch! Done
<!-- Reviewable comment -OQJ5lgi7wZGcuv8Oi67:bm4qzni -->
<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/cmd/roachtest/roachtestutil/profile.go:
```diff
@@ -304,3 +307,144 @@ func MeasureQPS(
+	}
+}
+
+// GetCPUProfile returns the CPU profile for the given node.
```
Done.
<!-- Reviewable comment -OQKWR8SEmPUN0TFxl9l:b-896fix -->
<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/cmd/roachtest/roachtestutil/profile.go:
```diff
@@ -304,3 +307,144 @@ func MeasureQPS(
+	client := DefaultHTTPClient(cluster, logger, HTTPTimeout(2*duration))
+	url := fmt.Sprintf("https://%s/debug/pprof/profile?seconds=%d", adminUIAddrs[0],
+		int(duration.Seconds()))
+	return getProfileWithTimeout(ctx, logger, client, url, 10*time.Second /* timeout */)
```
Makes sense. I increased the timeout to 15 seconds
<!-- Reviewable comment -OQJ6AGp-7OjTOr_AncT:b98l4ud -->
<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
Thanks for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146766#-:-OQJ5J9L6hWJvqHgm5tn:b-9qdoz3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @srosenberg, and @tbg)



<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachtest/roachtestutil/profile.go:
```diff
@@ -362,80 +363,73 @@ func getProfileWithTimeout(
 			defer profileWg.Done()
 			var err error
-			profiles[nodeID-1], err = collectProfile(nodeID)
+			profiles[nodeID-1], err = getProfileSingleNode(ctx, cluster, logger, profileType,
```
this probably works in practice with most callers, but shouldn't we really retain the original indexing of the input slice? Imagine a `option.NodeListOption{2,1}`. We'll put the profile for n2 into second position, but the caller will expect it in first.

iskettaneh replied:
That's true! Fixed it
<!-- Reviewable comment -OQhzf28017R_qnH69Ww:bvo6sek -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachtest/tests/sysbench.go:
```diff
@@ -596,6 +623,74 @@ func exportSysbenchResults(
+	var allocsProfiles []*profile.Profile
+	var mutexProfiles []*profile.Profile
+
+	m := t.NewGroup(task.WithContext(ctx))
```
nit: the caller has a task group, you could just pass it in. I think this can behave marginally better. For example, if the workload doesn't actually start (the one at the caller), you'll take a profile (of the idle cluster) and only then fail on the caller's `m.Wait`.

tbg replied:
For coherence, it would also be good to pass in the duration (`30*time.Second`) since these durations only make sense knowing the timing of the caller's workload.

iskettaneh replied:
I addressed the comment.
<!-- Reviewable comment -OQiHT42BbfI2okKRMsi:b-5vk2h -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachtest/tests/sysbench.go:
```diff
@@ -596,6 +623,74 @@ func exportSysbenchResults(
+
+	m.Wait()
+
+	// Merge the CPU profiles into one profile that contains all the nodes.
```
you can do similar `for` loop condensation here and below.

iskettaneh replied:
Done.
<!-- Reviewable comment -OQiHYpA72J92ghPeqfy:b-896fix -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachtest/tests/sysbench.go:
```diff
@@ -596,6 +623,74 @@ func exportSysbenchResults(
+	var mutexProfiles []*profile.Profile
+
+	m := t.NewGroup(task.WithContext(ctx))
+	m.Go(
```
(I think as of newer Go versions the `typ` variable is assigned locally for each loop, i.e. no need to write `typ := typ` https://go.dev/play/p/HJ9IV2AvWv_B)

iskettaneh replied:
Done.
<!-- Reviewable comment -OQi2hy9DFXl8uGE3O2D:b-896fix -->
<!-- Sent from Reviewable.io -->

tbg reviewed (changes requested):
Almost there!
I noticed there were a couple of force pushes, but your commit history looks like it's trying to be append-only. If you have muscle memory for rebasing against master, maybe this is helpful:

```
alias grb='git rebase -i $(git rev-list --merges -n 1 HEAD~)'
```

(just type `grb`).

This rebases on the branch point from whatever (release or master) branch you started your PR, so it will only rewrite the history starting at the first commit where you actually rearrange things (like when you're squashing some stuff together to make a new commit that you want to push).

iskettaneh committed:
roachtestutil: preserve the nodeList order when getting profiles

Release note: None

Epic: None

iskettaneh committed:
roachtest: clean up the function collectSysbenchProfiles

Release note: None

Epic: None

iskettaneh reviewed (commented):
Thanks for the review Tobi.

I addressed the comments. I haven't tested it yet (I am running gossip scale test). However, I won't merge it before I test it manually and make sure it's working as expected

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146766#-:-OQiMXwF4E1wT239hzxx:bmp6vlw)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @srosenberg, and @tbg)



<!-- Sent from Reviewable.io -->

herkolategan commented on pkg/cmd/roachtest/roachtestutil/profile.go:
```diff
@@ -304,3 +307,138 @@ func MeasureQPS(
+}
+
+// GetProfile collect profiles for all the nodes received in the function
+// parameters. It returns a slice of profiles and an error if any of the
```
The comment here indicates that it returns profiles and errors, but from the code below it looks like it's one or the other rather?

iskettaneh replied:
Done.
<!-- Reviewable comment -OQnOcwU3MYoHHgU3ZT1:b-896fix -->
<!-- Sent from Reviewable.io -->

herkolategan commented on pkg/cmd/roachtest/tests/sysbench.go:
```diff
@@ -596,6 +622,56 @@ func exportSysbenchResults(
+// collectSysbenchProfiles collects cpu, mutex, and allocs profiles from the
+// cluster `c` for the duration specified.
+func collectSysbenchProfiles(t test.Test, m task.Group, c cluster.Cluster, duration time.Duration) {
+	profiles := map[string][]*profile.Profile{"cpu": {}, "allocs": {}, "mutex": {}}
```
Readability Nit: Could be helpful to have the profile type as an actual type, like ProfileType that has convenience methods for also getting the endpoint name for the HTTP calls in:

```
	var actualProfileType string
	switch profileType {
	case "cpu":
		// It's unfortunate that the CPU profile endpoint is called "profile".
		actualProfileType = "profile"
	case "allocs":
```

tbg replied:
The pprof package also doesn't have such types, and (perhaps intentionally) keeps the set fluid. I am satisfied with the status quo, and would rather keep this as is to allow Ibrahim to get this PR in. (I'm not against adding this yourself later if you feel strongly).

iskettaneh replied:
ack
<!-- Reviewable comment -OQnQPMh-L0cquEjxE-e:b2nhysk -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachtest/roachtestutil/profile.go:
```diff
@@ -417,19 +417,19 @@ func GetProfile(
+	for i, nodeId := range nodes {
 		profileWg.Add(1)
-		go func(nodeID int) {
+		go func(nodeID int, sliceIdx int) {
```
no need to change it (!), but because you're using `sliceIdx` as the param name for the closure but `i` outside, there is now an extra variable that could mistakenly be used. This would be a little clearer if `i` were used for both.

tbg replied:
Additionally, having the order of the parameters be the reverse of the `range` return is also just a little off. It's obviously correct here, but the other order seems more canonical: `i, nodeID`.

iskettaneh replied:
I am now using i for both and I reversed the order to `i, nodeID`.
<!-- Reviewable comment -OQnPEu5192uGoCIRT9e:b-kk164f -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachtest/tests/sysbench.go:
```diff
@@ -259,6 +262,30 @@ func runSysbench(ctx context.Context, t test.Test, c cluster.Cluster, opts sysbe
+						opts.cmd(useHAProxy)+" run")
+
+					if msg, crashed := detectSysbenchCrash(result); crashed {
+						t.Skipf("%s; skipping test", msg)
```
actually let's not skip the test here. Just log this and return nil. We don't want this post-run collection to sacrifice a datapoint for a nightly run.

iskettaneh replied:
Done.
<!-- Reviewable comment -OQnQG9R3fs7JsevUNfE:b-896fix -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachtest/tests/sysbench.go:
```diff
@@ -282,8 +282,7 @@ func runSysbench(ctx context.Context, t test.Test, c cluster.Cluster, opts sysbe
 			time.Sleep(30 * time.Second)
-			collectSysbenchProfiles(t, ctx, c)
-			m.Wait()
+			collectSysbenchProfiles(t, m, c, time.Second*30 /* duration */)
```
this is a really big nit, and you don't have to change it. But the unit comes last: `30*time.Second`

iskettaneh replied:
Done.
<!-- Reviewable comment -OQnPZvA2RJUYKmZs76J:b-896fix -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachtest/roachtestutil/profile.go:
```diff
@@ -304,3 +307,138 @@ func MeasureQPS(
+
+	var latestError error
+	for {
+		select {
```
Good point, we should add a `case <-ctx.Done()` here.

tbg reviewed (commented):
Looks good after @herkolategan's comments 👍🏽

iskettaneh committed:
roachtest: fix nits in GetProfile()

This commit fixes:
1) Use errGroup instead of WaitGroup in GetProfile().
2) Stop skipping the test if running the benchmark in the end fails.

Release note: None

Epic: None

iskettaneh commented on pkg/cmd/roachtest/roachtestutil/profile.go:
```diff
@@ -304,3 +307,138 @@ func MeasureQPS(
+	profileWg.Wait()
+
+	// Return an error if any of the profiles failed to collect.
+	for _, err := range errors {
```
Done.
<!-- Reviewable comment -OQnWAMC1nmOmMm9tyAC:b-896fix -->
<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146766)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @srosenberg, and @tbg)



<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachtest/roachtestutil/profile.go:
```diff
@@ -413,31 +416,30 @@ func GetProfile(
-	// profileWg is used to wait for all the profiles to be collected.
-	profileWg := sync.WaitGroup{}
+	// Create a cancelable context and errgroup for concurrent profile collection.
+	ctx, cancel := context.WithCancel(ctx)
```
You don't need this - `errgroup.WithContext` internally makes a cancellable context and cancels it when it sees an error.

iskettaneh replied:
Oh that's neat!
<!-- Reviewable comment -OQsqwFN-hnzZ5iehfUx:bfjwqvg -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachtest/roachtestutil/profile.go:
```diff
@@ -413,31 +416,30 @@ func GetProfile(
 			if err != nil {
-				logger.Printf("error getting profile for node %d: %s", nodeID, err)
-				errors[sliceIdx] = err
+				logger.Printf("error getting profile for node %d: %s", nodeId, err)
```
since we're here: why not `return errors.Wrapf(err, "getting profile for n%d", nodeId)`

iskettaneh replied:
Done.
<!-- Reviewable comment -OQsrBkU5c2BAEcVZ-iR:b-896fix -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachtest/tests/sysbench.go:
```diff
@@ -273,7 +273,8 @@ func runSysbench(ctx context.Context, t test.Test, c cluster.Cluster, opts sysbe
 
 					if msg, crashed := detectSysbenchCrash(result); crashed {
-						t.Skipf("%s; skipping test", msg)
+						t.L().Printf("%s; sysbench run to collect profiles failed", msg)
```
Annoyingly, if this fails, we're still going to collect a set of profiles - but from an idle cluster. This is not ideal. Could you add the following:

- return marker error instead of `nil` (`var errSysbenchCrashed = errors.New("sysbench crashed")`)
- create profilesDir _above_ the call to `collectSysbenchProfiles` (i.e. that method gets a targetDir parameter)
- when the result of the sysbench run is known `m.Wait()`, delete the profile directory if seeing the marker error (`errors.Is(err, errSysbenchCrashed)`)

I'm also noticing that `m.Wait()` is currently called in `collectSysbenchProfiles`. I understand why - we need to see the profiles to merge them - but it has a smell to it. Instead, split out a `mergeSysbenchProfiles` function from `collectSysbenchProfiles` and then call collect, followed by wait, followed by either merge or delete.
Actually, looking into `collectSysbenchProfiles`, you would probably just inline whatever is left after pulling out the merge part, since it's mostly a few calls to `m.Go`.

iskettaneh replied:
Yeah that was annoying me as well, but I thought I merge it and see how often does this test fail (if ever). But since we are already here, I agree with you that we can proactively fix it.

I addressed the comments slightly differently.

1) I kept `collectSysbenchProfiles` but added `mergeAndExportSysbenchProfiles`. So we have one function to collect the profiles, and one to merge and export them.
2) I am not returning a specific error marker if sysbench fails. I am basically treating any error in either sysbench or while collecting/exporting profiles as an indicator that we need to cleanup the profiles directory. I am doing this because other potential errors are from collecting/exporting the profiles, and that could leave unwanted corrupt files in the directory.
3) Only if everything is good, we then proceed to exporting and merging the profiles.
<!-- Reviewable comment -OQsrKb17U_3b_2j1tcm:b-hav8m0 -->
<!-- Sent from Reviewable.io -->

tbg replied:
Good idea, that's simpler :thumbsup: 
<!-- Reviewable comment -OQwZHyF787dv5az_d1d:b-v0zr58 -->
<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Sorry to keep this up - but just one more thing that seems substantial enough to fix.

iskettaneh committed:
roachtest: clean up the sysbench profiles directory if workload fails

Release note: None

Epic: None

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146766)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @srosenberg, and @tbg)



<!-- Sent from Reviewable.io -->

tbg reviewed (changes requested):
A few more comments. Mainly, the profiles aren't yet collected as perf artifacts yet.

> Reviewed 1 of 2 files at r8, 1 of 3 files at r10, 2 of 2 files at r13, all commit messages.
> Dismissed @herkolategan from 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146766#-:-OQw_FyKAgY_65CdPhNN:b-7jlfiq)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @iskettaneh, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/sysbench.go` line 268 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/146766#-OQwYvri7edz218k0H_x:-OQwYvri7edz218k0H_y:b-9um37z) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e15393220d7e7c3e070abfa6416b99ed7b844c/pkg/cmd/roachtest/tests/sysbench.go#L268)):*
> ```Go
> 		{
> 			// Create the profiles directory in the artifacts directory.
> 			profilesDir := filepath.Join(t.ArtifactsDir(), "profiles")
> ```

Please take another look at [this comment](https://github.com/cockroachdb/cockroach/pull/146766#discussion_r2090484324) - I want to go a little further than just creating artifacts, as artifacts are not preserved in the same way that performance artifacts are. You need to put this into `t.PerfArtifactsDir()` and adjust the collection script to pick up these files as well.
___
*[`pkg/cmd/roachtest/tests/sysbench.go` line 297 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/146766#-OQwZNNuAnbxvHavNVnH:-OQwZNNuAnbxvHavNVnI:bavk0zo) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e15393220d7e7c3e070abfa6416b99ed7b844c/pkg/cmd/roachtest/tests/sysbench.go#L297)):*
> ```Go
> 			// up the profiles directory and return the error to avoid leaving
> 			// potentially corrupt profiles in the directory.
> 			if err != nil {
> ```

no need to change, but just mentioning that there is a way to express this block together with the next as

```
if errors.CombineErrors(err, m.WaitE()) != nil {
   ...
}
```
___
*[`pkg/cmd/roachtest/tests/sysbench.go` line 661 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/146766#-OQwZtPQDcbss-B88m7p:-OQwZtPQDcbss-B88m7q:bkrk6fi) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e15393220d7e7c3e070abfa6416b99ed7b844c/pkg/cmd/roachtest/tests/sysbench.go#L661)):*
> ```Go
> // cluster `c` for the duration specified. It returns a map of profiles for each
> // type and for each node. It returns an error if any of the profiles couldn't
> // be collected.
> ```

But that's not true. The method spawns a bunch of goroutines. Access to the returned map will be racy until `m` has been waited on. I would just inline this method; these semantics are not useful.


<!-- Sent from Reviewable.io -->

iskettaneh committed:
roachtest: inline collectSysbenchProfiles

This commit inlines collectSysbenchProfiles() as it had an implicit
requirement of calling m.wait() before using its return value.

Release note: None

Epic: None

iskettaneh reviewed (commented):
Discussed offline, we decided to leave uploading it as perf artifacts outside this PR

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146766#-:-OQxbYLYDv67KLxjaIkz:bhehc9g)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @srosenberg, and @tbg)

___
*[`pkg/cmd/roachtest/tests/sysbench.go` line 297 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/146766#-OQwZNNuAnbxvHavNVnH:-OQxG2evDsMcuRgy3DdW:b2nhysk) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e15393220d7e7c3e070abfa6416b99ed7b844c/pkg/cmd/roachtest/tests/sysbench.go#L297)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

no need to change, but just mentioning that there is a way to express this block together with the next as

```
if errors.CombineErrors(err, m.WaitE()) != nil {
   ...
}
```
</blockquote></details>

ack
___
*[`pkg/cmd/roachtest/tests/sysbench.go` line 661 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/146766#-OQwZtPQDcbss-B88m7p:-OQxG-4f-HRS2Yfkm1k9:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e15393220d7e7c3e070abfa6416b99ed7b844c/pkg/cmd/roachtest/tests/sysbench.go#L661)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

But that's not true. The method spawns a bunch of goroutines. Access to the returned map will be racy until `m` has been waited on. I would just inline this method; these semantics are not useful.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

herkolategan commented:
Thanks @iskettaneh, changes I mentioned look good to me. I'll leave it to @tbg to give the final check + approval.

tbg reviewed (approved):
:shipit: 

> Reviewed 1 of 1 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146766#-:-OQxuCaNExQDoV5cldan:bizlbj6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, and @srosenberg)



<!-- Sent from Reviewable.io -->
